protocols/gossipsub/src/handler.rs

```diff
@@ -554,6 +389,92 @@ impl ConnectionHandler for Handler {
 
         Poll::Pending
     }
+}
+
+impl ConnectionHandler for Handler {
+    type InEvent = HandlerIn;
+    type OutEvent = HandlerEvent;
+    type Error = Void;
+    type InboundOpenInfo = ();
+    type InboundProtocol = either::Either<ProtocolConfig, DeniedUpgrade>;
+    type OutboundOpenInfo = ();
+    type OutboundProtocol = ProtocolConfig;
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        match self {
+            Handler::Enabled(handler) => {
+                SubstreamProtocol::new(either::Either::Left(handler.listen_protocol.clone()), ())
+            }
+            Handler::Disabled(_) => {
+                SubstreamProtocol::new(either::Either::Right(DeniedUpgrade), ())
+            }
+        }
+    }
+
+    fn on_behaviour_event(&mut self, message: HandlerIn) {
+        match self {
+            Handler::Enabled(handler) => match message {
+                HandlerIn::Message(m) => handler.send_queue.push(m),
+                HandlerIn::JoinedMesh => {
+                    handler.in_mesh = true;
+                }
+                HandlerIn::LeftMesh => {
+                    handler.in_mesh = false;
+                }
+            },
+            Handler::Disabled(_) => {
+                log::debug!("Handler is disabled. Dropping message {:?}", message);
+            }
+        }
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        match self {
+            Handler::Enabled(handler) => {
+                if handler.in_mesh {
+                    return KeepAlive::Yes;
+                }
+
+                if let Some(
+                    OutboundSubstreamState::PendingSend(_, _)
+                    | OutboundSubstreamState::PendingFlush(_),
+                ) = handler.outbound_substream
+                {
+                    return KeepAlive::Yes;
+                }
+
+                KeepAlive::Until(handler.last_io_activity + handler.idle_timeout)
+            }
+            Handler::Disabled(_) => KeepAlive::No,
```

Comment:
I think this way it's possible we don't send the `PeerKind::UnsupportedProtocol` "on time" unless there is some kind of guarantee that poll will be called first **(?)**.  This event is important since we might want to get rid of (ban, or otherwise prevent connections to) this peer if it doesn't support what we need

Reply:
A handler is always completed drained of its events first before we even ask whether or not it should be shut down.

Reply:
See https://github.com/libp2p/rust-libp2p/blob/06ea8cef7ef6d8527fa9fdb9fc23dec88abe2a7b/swarm/src/connection.rs#L216-L303.

Reply:
Adding to the above, we first "drain" (i.e. `poll`) the `ConnectionHandler` till it returns `PollPending`:

https://github.com/libp2p/rust-libp2p/blob/06ea8cef7ef6d8527fa9fdb9fc23dec88abe2a7b/swarm/src/connection.rs#L231-L247

Only then do we check the `connection_keep_alive` statsu:

https://github.com/libp2p/rust-libp2p/blob/06ea8cef7ef6d8527fa9fdb9fc23dec88abe2a7b/swarm/src/connection.rs#L284-L303